Remove mutating global state in node-exporter objects

paulfantom opened the pull request:
Experiment to encapsulate node-exporter jsonnet definitions and prevent mutating global state.

First step towards https://github.com/prometheus-operator/kube-prometheus/issues/566

brancz commented on jsonnet/kube-prometheus/node-exporter/node-exporter.libsonnet:
```diff
@@ -1,214 +1,217 @@
-    imageRepos+:: { nodeExporter: 'quay.io/prometheus/node-exporter' },
+  nodeExporter+: {
+    _config+:: {
+      namespace: $._config.namespace,  //FIXME(paulfantom): figure out how to pass this global config in a better way
```
I like to use the `error "must provide namespace"` approach, which fails to generate if the config is not set, and forces the caller to configure it.

brancz reviewed (commented):
Nice start!

paulfantom commented:
I think this is ready for the next pass of reviews (cc @brancz @metalmatze @kakkoyun). Changes are done:
- moved `kube-rbac-proxy/container.libsonnet` to `kube-rbac-proxy/containerMixin.libsonnet` for now. End goal is to remove this completely and replace with what is new `kube-rbac-proxy/container.libsonnet`
- complete rewrite of node-exporter.libsonnet
- fixed incorrect version number in `app.kubernetes.io/version` label

I still need to fix unit tests and manifest generation, but from local testing it seems that manifests generated from `example.jsonnet` didn't change.

paulfantom commented:
Rebased to squash commits in some meaningful way.

brancz commented on jsonnet/kube-prometheus/kube-prometheus.libsonnet:
```diff
@@ -17,6 +18,11 @@ local kubeRbacProxyContainer = import './kube-rbac-proxy/container.libsonnet';
+  nodeExporter: nodeExporter({
+    namespace: $._config.namespace,
+    version: '1.0.1',
+    imageRepo: 'quay.io/prometheus/node-exporter',
```
if we're already refactoring, I would suggest to just have this be plain `image`, we tried this kind of "abstraction" in many ways, but we've learned that the image should just be treated as an opaque string

paulfantom replied:
I thought about it and applied such approach in new kube-rbac-proxy function. However in case of node-exporter I left it as is due to `app.kubernetes.io/version` label. I can see 3 options of satisfying label and image string:
1) current solution - compose `image` inside a function and pass `version` directly to label.
2) pass `image` string to a function and extract `version` from that string
3) pass `image` and `version` as strings.
I think 3rd one is the most fool-proof, but at the same time not so clean. That said I don't have any preference.

brancz replied:
I prefer 3, along the lines of, "prefer duplication over the wrong abstraction".

paulfantom replied:
Tested the 3rd approach in https://github.com/prometheus-operator/kube-prometheus/pull/862/commits/47f81f46f71a8cb672978bd17242b673472d8193#diff-46fb519eb228dee2044be788ed466089f6652ca4ea439f6f401dd3b509d12dcaR29-R30 and it looks a bit better and cleaner so I will change it here too.

brancz reviewed (commented):
So clean :bath: :shower: 

I'm loving that there is almost 0 diff in the generated manifests.

lgtm :+1: 

I wonder whether we should do these refactorings into a branch and merge that branch when everything is done, or cut a release branch now, and just merge the refactorings into master. I would think the later is probably more effective in moving forward faster, but the first would be safer in case it gets abandoned. Your call I would say.

paulfantom commented:
> I wonder whether we should do these refactorings into a branch and merge that branch when everything is done, or cut a release branch now, and just merge the refactorings into master

We do have `release-0.7` cut recently so I don't think we need another one and merging to `master` should be fine.

paulfantom commented:
I added info to readme and changed how `image` is passed.

paulfantom commented:
And accidentally closed my PR :smile:

lilic commented on README.md:
```diff
@@ -18,9 +18,14 @@ Components included in this package:
 
+## Warning
+
+`master` branch is under heavy refactoring work. Please use `release-0.7` branch until code refactoring is complete and this information is removed.
```
The problem is a lot of our users already use master, and we encouraged them to do that in many releases we said this. I suscept this will cause problems, is there no way to avoid it?

lilic replied:
Sorry missed the above comments, all for moving fast but might be worth notifying people in a way before this is merged for them to switch from master to latest release? WDYT of cutting a new release and mentioning that in the release notes? Or how long will this state take you think?

paulfantom replied:
`release-0.7` is very close to where current `master` is so I don't think a new branch is needed (the diff is in https://github.com/prometheus-operator/kube-prometheus/compare/release-0.7). As for announcing it, it was just done in prometheus community call, this PR mentions it at the top of the readme, and after it is merged I will announce it on slack.

I hope this all will take about a week (depending on how fast reviews will be), there are already 3 pending PRs on top of this one and I have more code on my PC to submit in the next PRs.

brancz replied:
I'd say additionally to this statement we can announce it in all the channels on kubernetes slack that are related to kube-prometheus. That, plus a commitment to get it done quickly, sounds like a plan to me.

lilic replied:
Sounds good to me, feel free to ping us on getting the reviews quickly in to not disrupt master for folks, thanks!

lilic reviewed (approved):
Thanks! lgtm
